adam/visualization/positioning.py

```diff
@@ -272,34 +320,76 @@ def for_objects(
         return AdamObjectPositioningModel(objects_to_bounding_boxes)
 
     def forward(self):  # pylint: disable=arguments-differ
-        distance_penalty = sum(
-            self.distance_to_origin_penalty(box) for box in self.object_bounding_boxes
-        )
         collision_penalty = sum(
             self.collision_penalty(box1, box2)
             for (box1, box2) in combinations(self.object_bounding_boxes, 2)
         )
+        below_ground_penalty = sum(
+            self.below_ground_penalty(box) for box in self.object_bounding_boxes
+        )
+        weak_gravity_penalty = sum(
+            self.weak_gravity_penalty(box) for box in self.object_bounding_boxes
+        )
         print(
-            f"distance penalty: {distance_penalty}\ncollision penalty: {collision_penalty}"
+            f"collision penalty: {collision_penalty}"
+            f"\nout of bounds penalty: {below_ground_penalty}"
+            f"\ngravity penalty: {weak_gravity_penalty}"
         )
-        return distance_penalty + collision_penalty
+        return collision_penalty + below_ground_penalty + weak_gravity_penalty
 
-    def dump_object_positions(self) -> None:
+    def dump_object_positions(self, *, prefix: str = "") -> None:
         for (adam_object, bounding_box) in self.adam_object_to_bounding_box.items():
-            print(f"{adam_object.name} = {bounding_box.center.data}")
+            print(f"{prefix}{adam_object.name} = {bounding_box.center.data}")
+
+    def get_object_position(self, obj: AdamObject) -> torch.Tensor:
+        """
+        Retrieves the (center) position of an AdamObject contained in this model.
+        Args:
+            obj: AdamObject whose position is requested
+
+        Returns: (3,) tensor of the requested object's position.
+
+        Raises KeyError if an AdamObject not contained in this model is queried.
+        """
+        return self.adam_object_to_bounding_box[obj].center.data
 
 
-class DistanceFromOriginPenalty(nn.Module):  # type: ignore
+class BelowGroundPenalty(nn.Module):  # type: ignore
     """
-    Model that penalizes boxes that are distant from the origin.
+    Model that penalizes boxes lying outside of the scene (i.e. below the ground plane) or off-camera)
     """
 
+    def __init(self) -> None:  # pylint: disable=useless-super-delegation
+        super().__init__()
+
+    def forward(self, *inputs):  # pylint: disable=arguments-differ
+        bounding_box: AxisAlignedBoundingBox = inputs[0]
+        distance_above_ground = bounding_box.distance_from_lowest_corner_to_ground()
+        if distance_above_ground >= 0:
+            return 0
+        else:
+            return -distance_above_ground
+
+
+class WeakGravityPenalty(nn.Module):  # type: ignore
+    """
+    Model that penalizes boxes that are not resting on the ground.
+    """
+
+    # TODO: exempt birds from this constraint
+    # TODO: exempt things resting on top of other objects from this constraint
```

Comment:
I don't think they need to be exempted - we'll let the non-collision constraint hold them up (just like real-life).